app/src/Model/Table/ArticlesTable.php

```diff
@@ -18,4 +23,46 @@ public function initialize(array $config): void
     {
         $this->addBehavior('Timestamp');
     }
+
+    /**
+     * Add slug with beforeSave event.
+     *
+     * @param \Cake\Event\EventInterface $event beforeSave event
+     * @param \App\Model\Entity\Article $entity initial beforeSave entity
+     * @param \ArrayObject $options beforeSave options
+     * @return void
+     */
+    public function beforeSave(EventInterface $event, Entity $entity, ArrayObject $options)
+    {
+        if ($entity->isNew() && !$entity->slug) {
+            $uuidLength = 5;
+            $sluggedTitle = Text::slug($entity->title);
+            // trim slug to maximum length defined in schema
+            $hashedSlug = substr(
+                $sluggedTitle,
+                0,
+                191 - $uuidLength
+            ) . '-' . substr(Security::hash(Text::uuid()), 0, $uuidLength);
+            $entity->slug = $hashedSlug;
+        }
```

Comment:
From comment https://github.com/combimauri/tutorial-cakephp/pull/2#discussion_r1385512930

Thank you for the suggestion. I think it's a bit overdo with hash & uuid. 

Basically, we can prevent using CakePHP ORM Table buildRules method & perform the validation.
1. About buildRules, please refer https://book.cakephp.org/4/en/orm/validation.html#creating-unique-field-rules
    - One example reside in UsersTable::buildRules where the bake command will automatically generate the rules based on the unique constraint set in the DB https://github.com/combimauri/tutorial-cakephp/blob/7da05953b24bebe8668595694f66cdf494adcb1e/app/src/Model/Table/UsersTable.php#L83
3. About validation, we can add validation for the field `title` in validationDefault method. We can write a custom validation rule 
    - Refer https://book.cakephp.org/4/en/core-libraries/validation.html#using-custom-validation-rules
    - It's better to use closure approach since we've to convert the title to the slug format. For example:
        ```
            ->add('title', 'unique', [
                'rule' => function ($value, $context) {
                    $conditions = [
                        'slug' => substr(Text::slug($value), 0, 191)
                    ];

                    if ($context['newRecord'] === false) {
                        $conditions['id !='] = $context['data']['id'];
                    }

                    return !$this->exists($conditions);
                }
            ])
        ```

Please check.

Reply:
Please check this topic https://book.cakephp.org/4/en/orm/validation.html#validation-vs-application-rules 

I hope it can give you an idea about the validation & application rules.